<!-- path follower launch
    Copyright Â© 2018 Naoki Akai. All rights reserved.
 -->

<launch >
    <!-- arguments -->
    <arg name="path_file" default="$(find path_follower)/../../maps/nic6f/path.txt" />
    <arg name="use_simple_obstacle_avoidance" default="false" />
    <arg name="map_frame" default="/map" />
    <arg name="path_topic_name" default="/target_path" />
    <arg name="service_path_topic_name" default="/static_target_path" />
    <arg name="base_link_frame" default="/base_link" />
    <arg name="input_max_vel_topic_name" default="/max_vel" />
    <arg name="output_twist_topic_name" default="/twist_cmd_path_follow" />
    <arg name="look_ahead_dist" default="2.0" />
    <arg name="max_vel" default="1.0" />
    <arg name="kv" default="1.0" />
    <arg name="cmd_publish_hz" default="30.0" />
    <arg name="stop" default="true" />
    <arg name="iterate_following" default="false" />
    <arg name="use_nav_core_server" default="false" />
    <arg name="input_scan_topic_name" default="/scan" />
    <arg name="local_map_size_x" default="10.0" />
    <arg name="local_map_size_y" default="10.0" />
    <arg name="local_map_pixel_size" default="0.05" />
    <arg name="robot_width" default="1.0" />
    <arg name="predict_time" default="5.0" />
    <arg name="step_time" default="0.1" />
    <arg name="kv_avoid" default="0.8" />
    <arg name="w_range" default="0.4" />
    <arg name="use_emergency_stop" default="false" />
    <arg name="stop_x_size" default="0.5" />
    <arg name="stop_y_size" default="0.3" />
    <arg name="stop_time" default="3.0" />
    <arg name="back_time" default="5.0" />
    <arg name="back_speed" default="0.2" />

    <!-- path server -->
    <node name="path_server" pkg="path_follower" type="path_server" args="$(arg path_file)" unless="$(arg use_nav_core_server)">
        <param name="/map_frame" value="$(arg map_frame)" type="string" />
        <param name="/path_topic_name" value="$(arg path_topic_name)" type="string" />
        <param name="/service_path_topic_name" value="$(arg service_path_topic_name)" type="string" />
    </node >

    <!-- path follower -->
    <node name="path_follower" pkg="path_follower" type="path_follower" output="screen" >
        <remap from="$(arg output_twist_topic_name)" to="/twist_cmd_raw" if="$(arg use_simple_obstacle_avoidance)" />
        <param name="/map_frame" value="$(arg map_frame)" type="string" />
        <param name="/base_link_frame" value="$(arg base_link_frame)" type="string" />
        <param name="/path_topic_name" value="$(arg path_topic_name)" type="string" />
        <param name="/input_max_vel_topic_name" value="$(arg input_max_vel_topic_name)" type="string" />
        <param name="/output_twist_topic_name" value="$(arg output_twist_topic_name)" type="string" />
        <param name="/look_ahead_dist" value="$(arg look_ahead_dist)" type="double" />
        <param name="/max_vel" value="$(arg max_vel)" type="double" />
        <param name="/kv" value="$(arg kv)" type="double" />
        <param name="/cmd_publish_hz" value="$(arg cmd_publish_hz)" type="double" />
        <param name="/stop" value="$(arg stop)" type="bool" />
        <param name="/iterate_following" value="$(arg iterate_following)" type="string" />
        <param name="/use_nav_core_server" value="$(arg use_nav_core_server)" type="bool" />
    </node >

    <!-- obstacle avoidance -->
    <node name="simple_obstacle_avoidance" pkg="path_follower" type="simple_obstacle_avoidance" output="screen" if="$(arg use_simple_obstacle_avoidance)" >
        <param name="/input_twist_topic_name" value="/twist_cmd_raw" type="string" />
        <param name="/input_scan_topic_name" value="$(arg input_scan_topic_name)" type="string" />
        <param name="/output_twist_topic_name" value="$(arg output_twist_topic_name)" type="string" />
        <param name="/local_map_size_x" value="$(arg local_map_size_x)" type="double" />
        <param name="/local_map_size_y" value="$(arg local_map_size_y)" type="double" />
        <param name="/local_map_pixel_size" value="$(arg local_map_pixel_size)" type="double" />
        <param name="/robot_width" value="$(arg robot_width)" type="double" />
        <param name="/predict_time" value="$(arg predict_time)" type="double" />
        <param name="/step_time" value="$(arg step_time)" type="double" />
        <param name="/kv" value="$(arg kv_avoid)" type="double" />
        <param name="/w_range" value="$(arg w_range)" type="double" />
        <param name="/use_emergency_stop" value="$(arg use_emergency_stop)" type="bool" />
        <param name="/stop_x_size" value="$(arg stop_x_size)" type="double" />
        <param name="/stop_y_size" value="$(arg stop_y_size)" type="double" />
        <param name="/stop_time" value="$(arg stop_time)" type="double" />
        <param name="/back_time" value="$(arg back_time)" type="double" />
        <param name="/back_speed" value="$(arg back_speed)" type="double" />
        <param name="/cmd_publish_hz" value="$(arg cmd_publish_hz)" type="double" />
    </node >
</launch >
